### DISCUSSION TITLE
A few questions about the project

### DISCUSSION THREAD

**Discussion by Alex87a** [Piero24/TelegramBot-AmazonOffers#2] [General]
Good evening! I have some questions about the project:

1) First of all, have you ever encountered the 429 error called "TooManyRequests"? Since these are new channels (at least so it seems to me looking at the ones you inserted in Linktree), you too, like me, should have a very low limit. As a result, Amazon may limit requests. Have you dealt with this before?

2) Would it be possible to have the part of code that deals with image customization? I'm using a Scraper and would like to be able to do something similar.

Thanks in advance!

**Comment by Piero24** [comment]
Hi @Alex87a,

1. The problem `429 error called "TooManyRequests"` often occurs if you have a new channel. It can happen for various reasons. Here you can find an answer I had given some time ago on [stackoverflow](https://stackoverflow.com/a/76180604/8723227).

    In short, if your APIs are new as it seems to me you must first make a sale through your affiliate link for the APIs to be unlocked. The unlock will last for 30 days from the last sale through your link. So if you make a sale today and tomorrow you get access to the APIs, then you're good for 30 days. If in the next 30 days you don't make any sales through your link, then they will be blocked until you make a new sale. If instead you have already made a sale through your affiliate link and more than 24-48 hours have passed and you can view the products through the amazon [scratchpad](https://webservices.amazon.it/paapi5/scratchpad/index.html) but not with the bot, then there may be a problem with the parameter settings.

2. For the image customization part, you can find it [here](https://github.com/Piero24/TelegramBot-AmazonOffers/blob/dc7b184212f8229a9fc0dd07333676d732d01c9c/src/media/image_generator.py). Here it takes the dimensions of the image and performs the necessary operations to apply it to the background image which you can find [here](https://github.com/Piero24/TelegramBot-AmazonOffers/tree/dc7b184212f8229a9fc0dd07333676d732d01c9c/src/media/template). If you want a different background image, just change the file name to `backg-2.jpg` or `backg-3.jpg` with your preferred one. If instead you want to modify the position of the elements in the image, you need to change the values in [image_generator.py](https://github.com/Piero24/TelegramBot-AmazonOffers/blob/dc7b184212f8229a9fc0dd07333676d732d01c9c/src/media/image_generator.py). **Be careful that I use `Pillow==9.5.0` because newer versions have changes that do not make the code work correctly**.

**Comment by Alex87a** [comment]
Good morning Piero, thanks for the explanation. Regarding the first answer: 

> if your APIs are new as it seems to me you must first make a sale through your affiliate link for the APIs to be unlocked. The unlock will last for 30 days from the last sale through your link. So if you make a sale today and tomorrow you get access to the APIs, then you're good for 30 days. If in the next 30 days you don't make any sales through your link, then they will be blocked until you make a new sale.

In reality, I have been in the affiliate program for a few years now, having never actually generated significant earnings from it (I think that in 2 years I will have generated around 32 euros between bounties and affiliate links). From what I understood, according to [this](https://webservices.amazon.com/paapi5/documentation/troubleshooting/api-rates.html) page, the limits are not reset every 30 days, but after the FIRST 30 days of use of the API we switch to a system that provides as many requests (TPS – Transactions per second and TPD – Transactions per day) as there are sales, with the increase of TPD of 1 every 0.05 euros. So I was wondering how a small affiliate could generate so much revenue in such a short time. What do you think about that?
P.S. You can safely move this conversation to the "Discussions" area, since this is not a problem with the project. I apologize for making it here.

**Reply by Piero24** [reply]
### For the first part:
> In reality, I have been in the affiliate program for a few years now, having never actually generated significant earnings from it (I think that in 2 years I will have generated around 32 euros between bounties and affiliate links). From what I understood, according to [this](https://webservices.amazon.com/paapi5/documentation/troubleshooting/api-rates.html) page, the limits are not reset every 30 days, but after the FIRST 30 days of use of the API we switch to a system that provides as many requests (TPS – Transactions per second and TPD – Transactions per day) as there are sales, with the increase of TPD of 1 every 0.05 euros.

You are right! I have also been registered with the affiliate program for a few years now but have never used it much. In any case, I noticed that when 30 days pass without you having made any further sales, the counter returns to zero and consequently you cannot make further calls and for every call you make the 429 error is immediately returned until you make an additional sale via your link. At that point, after 24-48 hours you will again be given the opportunity to make requests via API and as you say, as sales increase the number of requests you can make will also increase.

As also indicated in the link you inserted:
> Note that your account will lose access to Product Advertising API 5.0 if it has not generated referring sales for a consecutive 30-day period. If you lose access to Product Advertising API 5.0, you can continue to use other product linking tools, such as Site Stripe and generate revenue. You will regain access to Product Advertising API 5.0 within two days after your referred sales are shipped.

### For the second part:
> So I was wondering how a small affiliate could generate so much revenue in such a short time. What do you think about that?

A new affiliate really struggles at the beginning. Because this large limit does not allow you to compete with other affiliates who already have a greater number of subscribers to their channels and who therefore can make many more requests and have different channels.

Basically you have 2 options:
1. You are lucky enough that someone you know needs to buy a product that allows you to have a high TPS and TDP and you ask them to use your affiliate link. (But as I said before, having this 30 day "problem" you must then manage to increase the number of sales significantly in these 30 days so as not to lower or lose the use of the API and therefore return to the starting point).
2. You have to start slowly. Initially few requests and then gradually increase and expand the number of channels as your TPS and TDP increase.
The problem in both cases, however, is advertising. You need to invest some money and try to advertise your channel, blog or other on the various social platforms to attract subscribers. And therefore increase the probability of making sales.

**NOTE:** A further problem is that I believe many use scraping techniques or different APIs like those of KEEPA. As they allow you to have more information on the products such as discount codes or historical minimum prices etc. Which in turn attract further customers by giving them "extra service". Except that, for example, these APIs have a monthly cost to use them and by starting to use them immediately instead of PAAPIs you would have a double expense to bear. Advertising and APIs. So I believe it is more convenient to start with the PAAPI and then as soon as you have a sufficient number of subscribers to support the expense, move on to the KEEPA API. (Maybe there are alternative methods to get this additional information that I don't know but at the moment unfortunately only these options come to mind). In any case, this last one is a secondary problem. Solving it can help you provide an additional service and have more subscribers but it does not guarantee you will have a high TPS and TDP right away.
